src/tools/schema2base.ts

```diff
@@ -38,51 +41,110 @@ export abstract class Schema2Base {
-    const schemas: Dictionary<Schema> = {};
-    manifest.allSchemas.forEach(schema => schemas[schema.name] = schema);
-    for (const particle of manifest.allParticles) {
-      for (const connection of particle.connections) {
-        const schema = connection.type.getEntitySchema();
-        const name = schema && schema.names && schema.names[0];
-        if (name && !(name in schemas)) {
-          schemas[name] = schema;
-        }
-      }
-    }
     if (Object.keys(schemas).length === 0) {
       console.warn(`No schemas found in '${src}'`);
       return;
     }
 
-    // Collect inline schema fields. These will be output first so they're defined
-    // prior to use in their containing entity classes.
+    const inlineSchemas = Schema2Base.collectInlineSchemas(schemas);
+
+    const outFile = fs.openSync(outPath, 'w');
+    fs.writeSync(outFile, this.fileHeader(outName));
+    for (const dict of [inlineSchemas, schemas]) {
+      for (const [name, schema] of Object.entries(dict)) {
+        fs.writeSync(outFile, this.entityClass(name, schema).replace(/ +\n/g, '\n'));
+      }
+    }
+    fs.writeSync(outFile, this.fileFooter());
+    fs.closeSync(outFile);
+  }
+
+  /**
+   * Collect inline schema fields. These will be output first so they're defined
+   * prior to use in their containing entity classes.
+   * @param schemas
+   */
+  private static collectInlineSchemas(schemas: Dictionary<Schema>): Dictionary<Schema> {
     const inlineSchemas: Dictionary<Schema> = {};
     for (const schema of Object.values(schemas)) {
       for (const [field, descriptor] of Object.entries(schema.fields)) {
         if (descriptor.kind === 'schema-reference' && descriptor.schema.kind === 'schema-inline') {
-          const name = this.inlineSchemaName(field, descriptor);
+          const name = Schema2Base.inlineSchemaName(field, descriptor);
           if (!(name in inlineSchemas)) {
             inlineSchemas[name] = descriptor.schema.model.getEntitySchema();
           }
-         }
+        }
       }
     }
+    return inlineSchemas;
+  }
 
-    const outFile = fs.openSync(outPath, 'w');
-    fs.writeSync(outFile, this.fileHeader(outName));
-    for (const dict of [inlineSchemas, schemas]) {
-      for (const [name, schema] of Object.entries(dict)) {
-        fs.writeSync(outFile, this.entityClass(name, schema).replace(/ +\n/g, '\n'));
+  /**
+   * Collects schemas from the manfiest.
+   *
+   * Includes schemas that have no names (anonymous) or schemas with multiple aliases
+   *
+   * @param manifest Manifest expended by loader.
+   * @return Dictionary<Schema> target schemas for code generation.
+   */
+  private static collectSchemas(manifest: Manifest): Dictionary<Schema> {
+    /** Helper function: produce a new name for a schema found in the manifest. */
+    const mangleDuplicateName = <T>(collection: Dictionary<T>, name: string): string => {
+      let candidate = name;
+      while (candidate in collection) candidate += '_';
```

Comment:
I don't understand why we have to do such things. Let's VC?

Reply:
Sure, let's VC.

Reply:
For git historians out there, the takeaways from conversations with Myk and Piotr were: 
- need to construct some sort of data structure to mirror the type lattice within particles
- type compatibility is only needed within particles. A single particles input and output types should agree, given their definition. But we need not unify types across particles. Why? Runtime takes care of that. 
- best follow the example in the original [GH issue](https://github.com/PolymerLabs/arcs/issues/3663#issuecomment-538465035) for guidance